fish-rust/src/wutil/wrealpath.rs

```diff
@@ -0,0 +1,70 @@
+use std::{ffi::OsStr, fs::canonicalize, os::unix::prelude::OsStrExt};
+
+use cxx::let_cxx_string;
+
+use crate::{
+    ffi::{str2wcstring, wcs2string},
+    wchar::{wstr, WString},
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+
+/// Wide character realpath. The last path component does not need to be valid. If an error occurs,
+/// `wrealpath()` returns `None`
+pub fn wrealpath(pathname: &wstr) -> Option<WString> {
+    if pathname.is_empty() {
+        return None;
+    }
+
+    let mut real_path = Vec::new();
```

Comment:
Instead of declaring this as mut and then assigning to it (and clearing when it hasn't been written to) can you just return this from the `match` block below?